skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp

```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
+  }
+
+  auto cause = converted->cause.get();
+  auto causeId = cause ? cause->formID : 0;
+  auto target = converted->target.get();
+  auto targetId = target ? target->formID : 0;
+  auto perkId = converted->perkId;
+  auto flag = converted->flag;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([cause, causeId, target,
+                                                targetId, perkId, flag] {
+    auto obj = JsValue::Object();
+
+    auto causeLocal = RE::TESForm::LookupByID(causeId);
+    auto targetLocal = RE::TESForm::LookupByID(targetId);
+    auto perkLocal = RE::TESForm::LookupByID(perkId);
+
+    if (causeLocal == nullptr || targetLocal == nullptr ||
+        perkLocal == nullptr || causeLocal != cause || targetLocal != target ||
+        perkLocal->formType != RE::FormType::Perk || flag < 0) {
+      return;
+    }
+
+    obj.SetProperty("cause", CreateObject("ObjectReference", causeLocal));
+    obj.SetProperty("target", CreateObject("ObjectReference", targetLocal));
+    obj.SetProperty("perk", CreateObject("Perk", perkLocal));
+    obj.SetProperty("flag", JsValue::Double(flag));
+
+    EventsApi::SendEvent("perkEntryRun", { JsValue::Undefined(), obj });
+  });
+
+  return RE::BSEventNotifyControl::kContinue;
+}
+
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
+  const RE::PositionPlayerEvent* event,
+  RE::BSTEventSource<RE::PositionPlayerEvent>* eventSource)
+{
+  if (event == nullptr) {
+    return RE::BSEventNotifyControl::kContinue;
+  }
+
+  auto type = event->type;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([type] {
+    auto obj = JsValue::Object();
+
+    obj.SetProperty("eventType", JsValue::Double(static_cast<int>(type)));
+
+    EventsApi::SendEvent("positionPlayer", { JsValue::Undefined(), obj });
+  });
+
+  return RE::BSEventNotifyControl::kContinue;
+}
+
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
+  const RE::BGSFootstepEvent* event,
+  RE::BSTEventSource<RE::BGSFootstepEvent>* eventSource)
+{
+  if (event == nullptr) {
+    return RE::BSEventNotifyControl::kContinue;
+  }
+
+  auto tag = event->tag;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([tag] {
+    auto obj = JsValue::Object();
+
+    obj.SetProperty("tag", JsValue::String(static_cast<std::string>(tag)));
+
+    EventsApi::SendEvent("footstep", { JsValue::Undefined(), obj });
+  });
+
+  return RE::BSEventNotifyControl::kContinue;
+}
+
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
+  const RE::TESSceneEvent* event,
+  RE::BSTEventSource<RE::TESSceneEvent>* eventSource)
+{
+  auto converted = reinterpret_cast<const TESEvents::TESSceneEvent*>(event);
+
+  if (!converted) {
+    return RE::BSEventNotifyControl::kContinue;
+  }
+
+  auto flag = converted->flag;
+  auto sceneId = converted->sceneId;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([flag, sceneId] {
+    auto obj = JsValue::Object();
+
+    auto sceneLocal = RE::TESForm::LookupByID(sceneId);
+
+    if (sceneLocal == nullptr) {
```

Comment:
also check sceneLocal == scene

Reply:
we haven't scene form in event, only sceneId, i'll add formId check